### ISSUE TITLE
Actor Class IDL/.did file generation from parent actor

### ISSUE THREAD

**Issue by Gekctek** [dfinity/motoko#4422]
Something that has been a pain has been not being able to automatically generate .did files for my actor classes that are dynamically created via another actor. I have a solution where I manually compile the actor class file and generate a .did file, then use didc to generate JS/TS types for my clients, but it has limitations.

Is it feasible to generate multiple files, or have the .did file of the parent actor class include the definitions for both.
I do see a possible issue with that where maybe someone references an actor class to get types, but not deploy the actor class, so they may not want the .did to include it? but in that case, it can be fixed by moving types out of the actor class definition

I dont really care how its done specifically but i think its important to have the client generation to include these actor classes and this would be the first step

**Comment by chenyan-dfinity** [comment]
This is probably a ticket for `dfx`? On the compiler side, as you mentioned, you can specify the actor class as the main file and get the did file. Another option is that when you spawn the child canister, you should be able to get the did file from the canister metadata. Both of these options can be done as a custom build script in dfx.

The JS/TS binding can be generated via `dfx generate` without using `didc` (They are sharing exactly the same code).

Moving to the `dfx` side, once you spawn the child canister, you can add this canister a remote canister: https://github.com/dfinity/sdk/blob/master/CHANGELOG.md#feat-remote-canister-support. And `dfx generate` should generate bindings for remote canisters as well. One thing I am not sure is that if the `candid` field in remote canister is required or optional. If it's required, we should submit a issue to the sdk repo and allow `dfx` is fetch candid from the metadata.

**Comment by Gekctek** [comment]
The issue is that its not one or a static amount of canisters. Its one interface for many dynamic canisters, so i dont know how  I would put that in dfx, especially with local testing. 
Is this not a pattern for the actor model, to have dynamic child canisters?
Maybe dfx could add something that is a config for actor classes vs a remote canister?

**Comment by chenyan-dfinity** [comment]
You are right that there is no concept of actor class in `dfx`. As a workaround, you can put any instantiated child canister as a remote canister. The binding should be the same anyway.

Also, for the frontend, you cannot really call an actor class, only instantiated child canisters. Then the question is how does the frontend know the canister ids? Can the child canisters all managed from the Motoko side, instead of at the frontend side?

**Comment by Gekctek** [comment]
I have a workaround with a custom script to generate the did/js files manually

The ids of the canisters are part of the data. For example I have X amount of 'teams', each team has a principal that can be retrieved, which is the canister id
If I have the canisters be a gateway to the child canisters, im afraid of creating a bottleneck or extra latency

So i have everything working right now, I just have had many of headaches working with actor classes/child canisters, which feels like a pattern that should be encouraged with the actor model. Its not just here, but things like documentation is much harder to find on things like actor classes/child canisters vs just static actors

I can put something in the dfx issues and see what they have to say

**Comment by Gekctek** [comment]
https://github.com/dfinity/sdk/discussions/3640

**Comment by chenyan-dfinity** [comment]
> each team has a principal that can be retrieved, which is the canister id

Where do you retrieve this? From some endpoint in the main canister? Well, my point is that for `dfx` to support child canisters, there needs to be a way to gather this information, i.e., a dynamic directed graph connecting parent/child canisters. If we have a way to fetch this graph, we can build better support for them.
